### PR TITLE
[WiP] [JENKINS-27302] Compatible with Workflow

### PR DESCRIPTION
**Pull request by recena**
[JENKINS-27302](https://issues.jenkins-ci.org/browse/JENKINS-27302)

```
node {
  git 'https://github.com/recena/maven-helloworld'
  sh "mvn clean clover2:setup test clover2:aggregate clover2:clover"
  step([$class: 'CloverPublisher', cloverReportDir: 'target/site/clover', cloverReportFileName: 'clover.xml'])
}

### CODE REVIEW

File: src/main/java/hudson/plugins/clover/CloverPublisher.java [hunk 1]
```diff
@@ -123,12 +127,21 @@ public void setFailingTarget(CoverageTarget failingTarget) {
     /**
      * Gets the directory where the Clover Report is stored for the given build.
      */
+    @Deprecated
```
Comment (lordofthejars) [inline src/main/java/hudson/plugins/clover/CloverPublisher.java]: I don't know if it has sense , but when I see a `Deprecated` annotation I like to see the alternative so I don't have to navigate through all class searching the new method.

File: src/main/java/hudson/plugins/clover/CloverPublisher.java [hunk 2]
```diff
@@ -224,6 +280,54 @@ public String invoke(File file, VirtualChannel virtualChannel) throws IOExceptio
...
+        final File cloverXmlReport = getCloverXmlReport(build);
+        if (cloverXmlReport.exists()) {
+            listener.getLogger().println("Publishing Clover coverage results...");
+            ProjectCoverage result = null;
+            try {
+                result = CloverCoverageParser.parse(cloverXmlReport, workspacePath);
+            } catch (IOException e) {
+                Util.displayIOException(e, listener);
+                e.printStackTrace(listener.fatalError("Unable to copy coverage from " + coverageReport + " to " + buildTarget));
+                build.setResult(Result.FAILURE);
```
Comment (lordofthejars) [inline src/main/java/hudson/plugins/clover/CloverPublisher.java]: Should we put build as failed in case of couldn't parse a coverage file? Maybe see, simply I am wondering if we could add an error message in the report and that's all. Or maybe configure if we want to make build fail in these cases. Also I am thinking that probably this is a very strange error so we can leave as is.

Comment (recena) [inline src/main/java/hudson/plugins/clover/CloverPublisher.java]: @lordofthejars This is the original code. Maybe make sense in a different PR.

File: src/main/java/hudson/plugins/clover/CloverPublisher.java (hunk 1 above)
Comment (recena) [inline src/main/java/hudson/plugins/clover/CloverPublisher.java]: True.

File: src/main/java/hudson/plugins/clover/CloverPublisher.java (hunk 2 above)
Comment (lordofthejars) [inline src/main/java/hudson/plugins/clover/CloverPublisher.java]: Yes sure, I only mention this because in my case the diff appears as new (and also for me all the code is new). I was only wondering this as an improvement but nothing blocking. But at the same time I would say that if we adopt this policy, then I think that for example JaCoCo or JUnit plugins should follow the same approach.

Comment (recena) [inline src/main/java/hudson/plugins/clover/CloverPublisher.java]: @lordofthejars The goal is described here [JENKINS-27302](https://issues.jenkins-ci.org/browse/JENKINS-27302)

File: src/main/java/hudson/plugins/clover/CloverPublisher.java [hunk 3]
```diff
@@ -175,14 +199,105 @@ public boolean perform(AbstractBuild<?, ?> build, Launcher launcher, BuildListen
         return true;
     }
 
+    @Override
+    public void perform(Run<?, ?> run, FilePath workspace, Launcher launcher, TaskListener listener) throws
```
Comment (amuniz) [inline src/main/java/hudson/plugins/clover/CloverPublisher.java]: You can call this method from the deprecated one and avoid to duplicate a lot of code.

File: src/main/java/hudson/plugins/clover/CloverPublisher.java [hunk 4]
```diff
@@ -175,14 +199,105 @@ public boolean perform(AbstractBuild<?, ?> build, Launcher launcher, BuildListen
...
+            run.setResult(Result.FAILURE);
+        }
+    }
+
     /**
      * Process the clover.xml from the build directory. The clover.xml must have been already copied to the build dir.
      *
+     * @deprecated see {@link #processCloverXml(Run, FilePath, TaskListener, FilePath, FilePath)}
      */
+    @Deprecated
```
Comment (amuniz) [inline src/main/java/hudson/plugins/clover/CloverPublisher.java]: It's a weird thing to deprecate a method and add new code on it. Perhaps I'm missing something.

File: src/main/java/hudson/plugins/clover/results/AbstractCloverMetrics.java [hunk 5]
```diff
@@ -228,7 +228,7 @@ public void setName(String name) {
         this.name = name;
     }
 
-    public AbstractBuild getOwner() {
+    public Run<?, ?> getOwner() {
```
Comment (amuniz) [inline src/main/java/hudson/plugins/clover/results/AbstractCloverMetrics.java]: Using `@WithBridgeMethods` would avoid possible binary incompatibilities.

**Comment by omehegan** [comment]
@jglick might have some comments as well.

File: src/main/java/hudson/plugins/clover/CloverPublisher.java (hunk 4 above)
Comment (recena) [inline src/main/java/hudson/plugins/clover/CloverPublisher.java]: @amuniz Safer if you are using 1.580 or upper

Comment (amuniz) [inline src/main/java/hudson/plugins/clover/CloverPublisher.java]: It's private, it must be removed instead of deprecated.

**Comment by mparfianowicz** [comment]
Hi, 
Sorry that you had to wait so long. I completely forgotten about this PR.

I just merged your changes into the master branch. I applied some corrections:
 * I removed unnecessary code duplication (methods were copy-pasted, I extracted the common parts)
 * I downgraded clover plugin to be based on jenkins 1.580.3 (note that the 1.580.1 is the minimum version required for the workflow plugin)
 * I downgraded workflow-aggregator plugin used in tests to 1.3 (which is compatible with jenkins 1.580.1)

Tests are passing. Please build the snapshot and test it locally and let me know it it's fine - I'll be happy to release new Clover version with these changes.

**Comment by jacobfike** [comment]
When will this get released?

**Comment by lordofthejars** [comment]
It will be great to relase together with jacoco

**Comment by mparfianowicz** [comment]
I usually release new Jenkins Clover Plugin versions together with new releases of the Clover. Next Clover release (4.2.0, see https://confluence.atlassian.com/display/CLOVER/Clover+Road+Map) is planned for next quarter. 

But in case it's urgent, I could try to find some time to do this earlier. Or you can do it on your own if you have sufficient permissions: https://confluence.atlassian.com/display/CLOVER/Clover+for+Jenkins+Developer+Guide

**Comment by mparfianowicz** [comment]
Update: I will release it today. Cheers!

**Comment by mparfianowicz** [comment]
Done! :) 

You can download the plugin from http://repo.jenkins-ci.org/releases/org/jenkins-ci/plugins/clover/4.6.0/. 

Note: update of the Clover version in http://updates.jenkins-ci.org/update-center.json may take up to 6 hours, you'll have to wait to see the new version in the plugin manager.

**Comment by jacobfike** [comment]
Awesome, thanks!
